### PR TITLE
Loudness related thing support

### PR DESCRIPTION
**Pull request by apple502j**
This adds LoudnessHandler, which is mostly copy-paste from scratch-audio. Also, adds `Project.fireGreaterThanTrigger` and `Trigger.getOption`/`Trigger.castOption` as well as `Trigger.executed`. In terms of blocks, it adds `loudness` and `when () > ()` trigger.

### CODE REVIEW

**Comment by PullJosh** [comment]
I see you've updated this so it can be merged! I'm excited to take a close look at this PR soon. :)

File: src/Loudness.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      let sum = 0;
+      for (let i = 0; i < this.micDataArray.length; i++) {
+        sum += Math.pow(this.micDataArray[i], 2);
+      }
+      let rms = Math.sqrt(sum / this.micDataArray.length);
+      if (this._lastValue) {
+        rms = Math.max(rms, this._lastValue * 0.6);
+      }
+      this._lastValue = rms;
+      rms *= 1.63;
```
Comment (adroitwhiz) [inline src/Loudness.js:49]: This seems to be missing a lot of explanatory comments from Scratch's implementation.

File: src/Project.js [hunk 2]
```diff
@@ -82,6 +97,12 @@ export default class Project {
   }
 
   step() {
+    if (this.loudnessHandler.loudness > this._prevLoudness) {
+      this.fireGreatherThanTrigger(Trigger.LOUDNESS_GREATER_THAN);
+    }
+    this._prevLoudness = this.loudnessHandler.loudness;
+    this.fireGreatherThanTrigger(Trigger.TIMER_GREATER_THAN);
```
Comment (adroitwhiz) [inline src/Project.js:104]: Scratch seems to implement "edge-activated hats" (what we call "greater-than triggers" here) a bit differently--every step, it evaluates the predicate of the hat (e.g. "is loudness greater than N? is timer greater than N?") which returns either a true or false value. It fires those hat blocks if the predicate evaluates to true and it evaluated to false on the previous step (the predicate is on a rising edge).

This seems like it'd require less per-trigger logic, and the predicate logic can be encapsulated more cleanly. Would that be worth doing?

**Review by adroitwhiz — CHANGES_REQUESTED** [review]
I have a few questions about the implementation, particularly the way that "greater than" triggers are implemented.

File: src/Sprite.js [hunk 3]
```diff
@@ -305,6 +304,10 @@ class SpriteBase {
   get answer() {
     return this._project.answer;
   }
+
+  async loudness() {
+    return this._project.loudnessHandler.getLoudness();
```
Comment (adroitwhiz) [inline src/Sprite.js:309]: It looks like Scratch [only recomputes loudness once per step](https://github.com/LLK/scratch-vm/blob/f405e59d01a8f9c0e3e986fb5276667a8a3c7d40/src/blocks/scratch3_sensing.js#L262-L275).

**Pull request closed without merging**